### PR TITLE
Optimise Minz_ModelPdo::class

### PR DESCRIPTION
**Pull request by ColonelMoutarde**
- Fix typo,
- remove unnecessary null in property,
- remove unused property,
- add phpDoc,
- add ext PDO in composer.json,
- use strict comparison,
- indentation

### CODE REVIEW

File: lib/Minz/ModelPdo.php [hunk 1]
```diff
@@ -6,22 +6,27 @@
  */
 
 /**
- * La classe Model_sql représente le modèle interragissant avec les bases de données
+ * La classe Model_sql représente le modèle interagissant avec les bases de données
```
Comment (Frenzie) [inline lib/Minz/ModelPdo.php]: If we're changing the line anyway, should we change it to English?

`The Model_sql class represents the database interaction model`, or something along those lines.

Comment (ColonelMoutarde) [reply lib/Minz/ModelPdo.php]: Done

File: lib/Minz/ModelPdo.php [hunk 2]
```diff
@@ -6,22 +6,27 @@
  */
 
 /**
- * La classe Model_sql représente le modèle interragissant avec les bases de données
+ * The Model_sql class represents the model interacting with the databases
```
Comment (Frenzie) [inline lib/Minz/ModelPdo.php]:
```suggestion
 * The Model_sql class represents the model for interacting with databases.
```

File: lib/Minz/ModelPdo.php [hunk 3]
```diff
@@ -68,22 +73,27 @@ private function dbConnect() {
...
 		if ($currentUser === null) {
 			$currentUser = Minz_Session::param('currentUser');
 		}
-		if ($currentPdo != null) {
+		if ($currentPdo !== null) {
 			$this->pdo = $currentPdo;
 			return;
 		}
-		if ($currentUser == '') {
+		if ($currentUser === '') {
```
Comment (Frenzie) [inline lib/Minz/ModelPdo.php]: I'm not sure if stricter matching is necessarily intended here if it was written like that.

**Review by Frenzie — COMMENTED** [review]
Looks fine to me in principle but @Alkarex should be more familiar with the rationale here. I'm less sure about the strict comparison changes for example.

File: lib/Minz/ModelPdo.php (hunk 3 above)
Comment (ColonelMoutarde) [reply lib/Minz/ModelPdo.php]: The strict comparison is always more qualitative than the non-strict comparison. The non-strict one has too much risk of side effects.
Look at the table in the PHP documentation
https://www.php.net/manual/en/types.comparisons.php

**Comment by Alkarex** [comment]
Let me merge https://github.com/FreshRSS/FreshRSS/pull/4110 first, to avoid more merge conflicts

File: lib/Minz/ModelPdo.php (hunk 3 above)
Comment (Alkarex) [reply lib/Minz/ModelPdo.php]: Changing to `===` would mean that you believe that `false`, `0`, `''`, are valid usernames, which is not the case.

Comment (ColonelMoutarde) [reply lib/Minz/ModelPdo.php]: If I set === and $currentUser is an empty string, then an exception is thrown if we don't want 0, false to be accepted, so let's make it explicit.

Comment (ColonelMoutarde) [reply lib/Minz/ModelPdo.php]: changed

**Comment by Alkarex** [comment]
As you can see, the automated tests are failing.
You can run some of them locally, e.g. with `composer test`

**Comment by Alkarex** [comment]
Related https://github.com/FreshRSS/FreshRSS/issues/4112

File: lib/Minz/ModelPdo.php (hunk 3 above)
Comment (Frenzie) [reply lib/Minz/ModelPdo.php]: Incidentally, my comment was not intended for just this one specific line but just as a representative example. They were all explicitly written to be truthy or falsy, otherwise they would've been strict already.

Comment (ColonelMoutarde) [reply lib/Minz/ModelPdo.php]: ;) https://stackoverflow.com/questions/2401478/why-is-faster-than-in-php

Comment (Frenzie) [reply lib/Minz/ModelPdo.php]: My point isn't that I like `($currentUser == '')` better than `($currentUser === '' || $currentUser === null || $currentUser === false || $currentUser === 0)`)[1] but that it entails all that and produces different results. I.e., that `($currentUser == '')` isn't the same as `($currentUser === '')`. That's the same for probably every instance where you replaced `!=` or `==` by the stricter `!==` or `===`. It's quite unlikely that truthy and falsy were used by mistake (in this particular codebase).

[1] Though for reading I probably do since it says all that so much quicker. :-) But that's neither here nor there.

File: lib/Minz/ModelPdo.php [hunk 4]
```diff
@@ -114,20 +127,43 @@ public function __construct($currentUser = null, $currentPdo = null) {
 			);
 	}
 
-	public function beginTransaction() {
+	/**
+	 * @return void
+	 */
+	public function beginTransaction()
+	{
```
Comment (Alkarex) [inline lib/Minz/ModelPdo.php]: Please keep same curly-brackets alignment than the rest of the document

File: lib/Minz/ModelPdo.php [hunk 5]
```diff
@@ -96,10 +109,10 @@ public function __construct($currentUser = null, $currentPdo = null) {
 			try {
 				$this->dbConnect();
 				return;
-			} catch (PDOException $e) {
+			} catch (Minz_PDOConnectionException $e) {
 				$ex = $e;
-				if (empty($e->errorInfo[0]) || $e->errorInfo[0] !== '08006') {
```
Comment (Alkarex) [inline lib/Minz/ModelPdo.php:101]: The filter on error info `08006` seems gone

File: lib/Minz/ModelPdo.php [hunk 6]
```diff
@@ -67,22 +75,27 @@ private function dbConnect() {
...
+	 * @throws Minz_PDOConnectionException
 	 */
-	public function __construct($currentUser = null, $currentPdo = null) {
+	public function __construct($currentUser = null, $currentPdo = null)
+	{
 		if ($currentUser === null) {
 			$currentUser = Minz_Session::param('currentUser');
 		}
-		if ($currentPdo != null) {
+		if ($currentPdo !== null) {
```
Comment (Frenzie) [inline lib/Minz/ModelPdo.php:102]: You reverted the one, but as I said I was talking about all of them. This particular one does feel a little suspicious in the original in the sense that using null instead of '' is a bit unidiomatic (in this codebase) but when I read this, I very much do read it as an intentional $currentPdo isn't an empty string (or however unlikely 0 or false) either.

File: lib/Minz/ModelPdo.php [hunk 7]
```diff
@@ -96,10 +109,10 @@ public function __construct($currentUser = null, $currentPdo = null) {
 			try {
 				$this->dbConnect();
 				return;
-			} catch (PDOException $e) {
+			} catch (Minz_PDOConnectionException $e) {
```
Comment (Alkarex) [inline lib/Minz/ModelPdo.php]: We also need to catch other PDO exceptions

**Review by Alkarex — CHANGES_REQUESTED** [review]
See comments

File: lib/Minz/ModelPdo.php [hunk 8]
```diff
@@ -67,22 +75,27 @@ private function dbConnect() {
...
+		if ($currentPdo !== null) {
 			$this->pdo = $currentPdo;
 			return;
 		}
 		if ($currentUser == '') {
 			throw new Minz_PDOConnectionException('Current user must not be empty!', '', Minz_Exception::ERROR);
 		}
-		if (self::$usesSharedPdo && self::$sharedPdo != null &&
-			($currentUser == '' || $currentUser === self::$sharedCurrentUser)) {
+		if (self::$usesSharedPdo && self::$sharedPdo !== null &&
```
Comment (Frenzie) [inline lib/Minz/ModelPdo.php]: Also these.

**Review by Frenzie — COMMENTED** [review]
(I'm not saying I'm against changing these lines, but in the diff it looks like a seemingly minor aesthetic swap with grave consequences in logic.)

File: lib/Minz/ModelPdo.php (hunk 6 above)
Comment (Alkarex) [reply lib/Minz/ModelPdo.php:102]: Although I believe that the original code is safer for now, the new type hints will mitigate the issue by warning us if we have code risking calling with something that is not of type `PDO|null`

File: lib/Minz/ModelPdo.php (hunk 4 above)
Comment (ColonelMoutarde) [reply lib/Minz/ModelPdo.php]: ok, your not PSR complient ?

File: lib/Minz/ModelPdo.php (hunk 8 above)
Comment (ColonelMoutarde) [reply lib/Minz/ModelPdo.php]: Go to your mattermost ?

Comment (ColonelMoutarde) [reply lib/Minz/ModelPdo.php]: https://framateam.org

File: lib/Minz/ModelPdo.php (hunk 7 above)
Comment (ColonelMoutarde) [reply lib/Minz/ModelPdo.php]: This case is tagged to "ok" by phpstan PDOException not

File: lib/Minz/ModelPdo.php (hunk 6 above)
Comment (Frenzie) [reply lib/Minz/ModelPdo.php:102]: I see, that's a bit of rationale missing in the commit message that I overlooked. To what extent can these type hints be trusted though?

File: lib/Minz/ModelPdo.php (hunk 4 above)
Comment (Alkarex) [reply lib/Minz/ModelPdo.php]: See https://github.com/FreshRSS/FreshRSS/blob/edge/phpcs.xml
In general, please keep the convention used in a given file

**Comment by Alkarex** [comment]
Slightly related, I have just updated https://github.com/FreshRSS/FreshRSS/pull/4121 to be able to run tests in PHP 8.1, which also shows that we have a number of PHP 8.1+ warnings to address

File: lib/Minz/ModelPdo.php (hunk 6 above)
Comment (Alkarex) [reply lib/Minz/ModelPdo.php:102]: > To what extent can these type hints be trusted though?

Only when we run PHPStan for now, but that is a part of the CI. But Minz (in theory) can also be used outside this context.

File: lib/Minz/ModelPdo.php (hunk 7 above)
Comment (Alkarex) [reply lib/Minz/ModelPdo.php]: Maybe, but that probably hides another problem, because this is wrong

File: lib/Minz/ModelPdo.php (hunk 4 above)
Comment (Alkarex) [reply lib/Minz/ModelPdo.php]: @Frenzie If you have a minute, could you please check why this not caught by PHPCS?

Comment (Frenzie) [reply lib/Minz/ModelPdo.php]: At a glance it looks like we only demand a newline after an opening brace and I guess no one's ever tried to use this style before. Probably easy to work around, let me see if I can figure it out in roughly 5 minutes (since I should sleep soon :-) ).

Comment (Frenzie) [reply lib/Minz/ModelPdo.php]: https://github.com/FreshRSS/FreshRSS/pull/4122

**Review by Alkarex — APPROVED** [review]

File: composer.json [hunk 9]
```diff
@@ -17,7 +17,9 @@
         "WebSub"
     ],
     "require": {
-        "php": ">=7.0.0"
+        "php": ">=7.0.0",
+        "ext-pdo": "*",
+        "ext-simplexml": "*"
```
Comment (Alkarex) [inline composer.json]: This could better be in another PR, but that does not matter, and the list could be extended:
https://github.com/FreshRSS/FreshRSS/blob/8e398d24f132095fffba0db3c7ceea9a43371b4d/Docker/Dockerfile-Alpine#L7-L9

**Comment by Alkarex** [comment]
@ColonelMoutarde I have fixed the Exception issue along with other minor changes. You are welcome to have a look

**Comment by Alkarex** [comment]
https://github.com/FreshRSS/FreshRSS/pull/4140
